common/channel/src/libra_channel.rs

```diff
@@ -157,9 +173,8 @@ impl<K: Eq + Hash + Clone, M> Stream for Receiver<K, M> {
                 let _err = status_ch.send(ElementStatus::Dequeued);
             }
             Poll::Ready(Some(val))
-        // if the only Arc reference to `shared_state` is 1, which is this receiver,
-        // this must mean all senders have been dropped (and so the stream is terminated)
-        } else if Arc::strong_count(&self.shared_state) == 1 {
+        // all senders have been dropped (and so the stream is terminated)
+        } else if shared_state.num_senders == 0 {
```

Comment:
I see, so now we're using the count rather than the count of references.

I'm new to the asynchronus model of Rust, but should we use an atomic type?

Also, I would just for safety, even though I know you have a ton off asserts around it, to say `shared_state.num_senders <= 0`

Reply:
1. `num_senders` is inside an `Arc<Mutex<SharedState>>`, so we can only read/write it if we hold the lock. Rust (mostly) inherits the C++ atomic memory model. We could probably get away with `num_senders: AtomicUsize` outside the mutex (so the shared part would be `Arc<(AtomicUsize, Mutex<SharedState>)>`), then use `Ordering::AcqRel` to increment/decrement the counter before acquiring the lock in `drop`, and `Ordering::Acquire` in the `Receiver` to observe `num_senders`. However, I would only be comfortable adding this if we use a concurrency model checker like [tokio-rs/loom](https://github.com/tokio-rs/loom). Unfortunately, setting up loom in libra is somewhat painful and I don't really have the cycles to spare on it : (

2. `num_senders: usize` so it's already implied that `num_senders < 0` is impossible : )

Reply:
Oh, duh `usize`